### PR TITLE
[ios][audio] Add audio metadata (title) updates

### PR DESCRIPTION
**Pull request by dani-mp**
⚠️ This is my first code contribution to the project. I'm aware my implementation is not yet complete. I would appreciate it if folks can guide me so we can land the feature in the best way and as soon as possible.

# Why

With this change, an iOS user can opt-in to receive metadata—only the title, for now—updates while playing a sound if any.

[Related feature request.](https://expo.canny.io/feature-requests/p/event-support-for-fetching-video-metadata)

<details>
<summary>Screenshot from bare-expo app player</summary>

Simulator Screen Shot - iPhone 12 - 2021-08-20 at 16 29 56

</details>

# How

On the JavaScript side, I've added a new method to the Sound class, `setOnMetadataUpdate`, which receives a callback that is executed when the metadata of the currently playing audio changes.

On the iOS side, I'm KVOing the `timedMetadata` property of the player's `currentItem`, and grab the relevant info before emitting a `didUpdateMetadata` event.

For now, a metadata payload is an object with the title—if any—of the song, using AVFoundation's `AVMetadataCommonKeyTitle`.

# Test Plan

Go to the audio player in the bare expo app and play an audio stream that will eventually provide a title in its metadata. That title should appear in the UI, under the play/pause button.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: apps/native-component-list/src/screens/AV/AudioPlayer.tsx [hunk 1]
```diff
@@ -22,6 +22,9 @@ interface Props {
   source: PlaybackSource;
 }
 
+// TODO: Grab the type from expo-av.
```
Comment (dani-mp) [inline apps/native-component-list/src/screens/AV/AudioPlayer.tsx:25]: How can I do this? I've declared the type in `AV.ts` but I can't access it from `expo-av`.

File: packages/expo-av/CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,7 @@
 - Add web support for recording. ([#8721](https://github.com/expo/expo/pull/8721) by [@WazzaJB](https://github.com/WazzaJB) and [@mnightingale](https://github.com/mnightingale))
 - Add permissions support for web. ([#8721](https://github.com/expo/expo/pull/8721) by [@mnightingale](https://github.com/mnightingale))
 - Add Audio `usePermissions` hook from modules factory. ([#13851](https://github.com/expo/expo/pull/13851) by [@bycedric](https://github.com/bycedric))
+- Add Audio metadata (title) updates for iOS via `setOnMetadataUpdate`. (by [@dani-mp](https://github.com/dani-mp)) ([#14134](https://github.com/expo/expo/pull/14134) by [@dani-mp](https://github.com/dani-mp))
```
Comment (dani-mp) [inline packages/expo-av/CHANGELOG.md:15]:
```suggestion
- Add Audio metadata (title) updates for iOS via `setOnMetadataUpdate`. ([#14134](https://github.com/expo/expo/pull/14134) by [@dani-mp](https://github.com/dani-mp))
```

File: apps/native-component-list/src/screens/AV/AudioPlayer.tsx (hunk 1 above)
Comment (bbarthec) [reply apps/native-component-list/src/screens/AV/AudioPlayer.tsx:25]: I see that there's a problem with exporting the types from this module.
As you can see in [`package.json`](https://github.com/expo/expo/blob/master/packages/expo-av/package.json) the entry point for the types for this module is [`build/index.d.ts`](https://github.com/expo/expo/blob/master/packages/expo-av/build/index.d.ts) that comes from [`src/index.ts`](https://github.com/expo/expo/blob/master/packages/expo-av/src/index.ts). This latter file does not re-export the exported values nor types from [`src/AV.ts`](https://github.com/expo/expo/blob/master/packages/expo-av/src/AV.ts) directly. Unfortunately, I don't know why it's written that way (😅), but I believe it should be fixed to export all the that are exposed in the module's API 💪  
I suggest adding the following line to the [`src/index.ts`](https://github.com/expo/expo/blob/master/packages/expo-av/src/index.ts) file, rebuild the package and you'll be able to use declared in [`src/AV.ts`](https://github.com/expo/expo/blob/master/packages/expo-av/src/AV.ts) type 🥳 
```diff
  import * as Audio from './Audio';
  export { Audio };
  export {
    default as Video,
    VideoProps,
    VideoState,
    AVPlaybackNativeSource,
    AVPlaybackStatus,
    AVPlaybackStatusToSet,
    VideoNativeProps,
    VideoNaturalSize,
    VideoReadyForDisplayEvent,
    ResizeMode,
    VideoFullscreenUpdateEvent,
  } from './Video';
+ export { AVPlaybackMetadata } from './AV';
```

File: apps/native-component-list/src/screens/AV/Player.tsx [hunk 3]
```diff
@@ -175,6 +176,8 @@ export default function Player(props: Props) {
         {_renderReplayButton()}
       </View>
 
+      <Text>{props.metadata.title || ''}</Text>
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/AV/Player.tsx:179]: I guess this is more semantically correct 😅 
```suggestion
      <Text>{props.metadata.title ?? ''}</Text>
```

**Review by bbarthec — CHANGES_REQUESTED** [review]
@dani-mp, thank you very much fir your first contribution 🥳 You've followed the contributors guide excellently 💪 
My remarks:
- pay attention to indentation (we use `2 spaces` everywhere in the code)
- regarding the misexported type I've provided some insights in the comments
- I need some documentation (obligatory) and possibly Android/Web implementation (optional) as well 👀

File: packages/expo-av/ios/EXAV/EXAVPlayerData.m [hunk 4]
```diff
@@ -798,6 +802,17 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
...
+            NSArray<AVMetadataItem *> *timedMetadata = strongSelf.player.currentItem.timedMetadata;
+            NSMutableDictionary *metadata = [@{} mutableCopy];
+            for (AVMetadataItem *item in timedMetadata) {
+                if ([item.commonKey isEqual:AVMetadataCommonKeyTitle]) {
+                    NSString *title = item.stringValue;
+                    [metadata setObject:title forKey:@"title"];
+                    break;
+                }
+            }
+            strongSelf.metadataUpdateCallback(metadata);
```
[the remark this answers is not in the document]
Comment (dani-mp) [reply packages/expo-av/ios/EXAV/EXAVPlayerData.m]: Changed. Sorry about that!

Small complaint: without an Objective C linter and 4 spaces being the default in Xcode, this requirement is not a great contributing experience. 🙇

File: packages/expo-av/src/Audio/Sound.ts [hunk 5]
```diff
@@ -91,6 +93,18 @@ export class Sound implements Playback {
...
 
+  _internalMetadataUpdateCallback = ({
+    key,
+    metadata,
+  }: {
+    key: AudioInstance;
+    metadata: AVPlaybackMetadata;
+  }) => {
+    if (this._key === key && this._onMetadataUpdate) {
+      this._onMetadataUpdate(metadata);
```
[the remark this answers is not in the document]
Comment (dani-mp) [reply packages/expo-av/src/Audio/Sound.ts]: Changed!

File: packages/expo-av/ios/EXAV/EXAV.m [hunk 6]
```diff
@@ -650,6 +652,14 @@ - (void)_removeAudioRecorder:(BOOL)removeFilenameAndSettings
     }
   };
   
+    data.metadataUpdateCallback = ^(NSDictionary *metadata) {
+      EX_ENSURE_STRONGIFY(self);
+        if (self.isBeingObserved) {
+          NSDictionary<NSString *, id> *response = @{@"key": key, @"metadata": metadata};
+          [self sendEventWithName:EXDidUpdateMetadataEventName body:response];
+        }
```
[the remark this answers is not in the document]
Comment (dani-mp) [reply packages/expo-av/ios/EXAV/EXAV.m]: Changed!

File: apps/native-component-list/src/screens/AV/Player.tsx (hunk 3 above)
Comment (dani-mp) [reply apps/native-component-list/src/screens/AV/Player.tsx]: Changed!

**Comment by dani-mp** [comment]
> * I need some documentation (obligatory) and possibly Android/Web implementation (optional) as well 👀

I'll write some docs tomorrow for this. About the Android/Web implementation, do you think we can do it in a different PR? Especially because I'm not that familiar with those APIs I wouldn't want to block this PR.

**Comment by dani-mp** [comment]
@bbarthec I've added docs! Let me know what you think and if we can move forward with the PR!

Thanks a lot for your help.

File: packages/expo-av/CHANGELOG.md [hunk 7]
```diff
@@ -12,6 +12,7 @@
 - Add web support for recording. ([#8721](https://github.com/expo/expo/pull/8721) by [@WazzaJB](https://github.com/WazzaJB) and [@mnightingale](https://github.com/mnightingale))
 - Add permissions support for web. ([#8721](https://github.com/expo/expo/pull/8721) by [@mnightingale](https://github.com/mnightingale))
 - Add Audio `usePermissions` hook from modules factory. ([#13851](https://github.com/expo/expo/pull/13851) by [@bycedric](https://github.com/bycedric))
+- Add Audio metadata (title) updates for iOS via `setOnMetadataUpdate`. ([#14134](https://github.com/expo/expo/pull/14134) by [@dani-mp](https://github.com/dani-mp))
```
Comment (bbarthec) [inline packages/expo-av/CHANGELOG.md:15]:
```suggestion
- On iOS add Audio metadata (title) updates via `setOnMetadataUpdate`. ([#14134](https://github.com/expo/expo/pull/14134) by [@dani-mp](https://github.com/dani-mp))
```

**Review by bbarthec — COMMENTED** [review]
It looks good now 👍 
@dani-mp thank you very much for your contribution here 🙇 🎉 
Documentation looks fine to me 👀 
Fix the conflicts (by rebasing with the master and rebuilding the package) and we're ready to merge 🏃

File: packages/expo-av/CHANGELOG.md (hunk 7 above)
Comment (dani-mp) [inline packages/expo-av/CHANGELOG.md:15]:
```suggestion
- On iOS, add Audio metadata (title) updates for iOS via `setOnMetadataUpdate`. ([#14134](https://github.com/expo/expo/pull/14134) by [@dani-mp](https://github.com/dani-mp))
```

File: packages/expo-av/CHANGELOG.md [hunk 8]
```diff
@@ -12,6 +12,7 @@
 - Add web support for recording. ([#8721](https://github.com/expo/expo/pull/8721) by [@WazzaJB](https://github.com/WazzaJB) and [@mnightingale](https://github.com/mnightingale))
 - Add permissions support for web. ([#8721](https://github.com/expo/expo/pull/8721) by [@mnightingale](https://github.com/mnightingale))
 - Add Audio `usePermissions` hook from modules factory. ([#13851](https://github.com/expo/expo/pull/13851) by [@bycedric](https://github.com/bycedric))
+- On iOS, add Audio metadata (title) updates for iOS via `setOnMetadataUpdate`. ([#14134](https://github.com/expo/expo/pull/14134) by [@dani-mp](https://github.com/dani-mp))
```
Comment (bbarthec) [inline packages/expo-av/CHANGELOG.md:15]: I'm sorry I'm suggesting another update here 😓 
I've, for some reason, failed to remove the original `iOS` word position 😅 
```suggestion
- On iOS, add Audio metadata (title) updates via `setOnMetadataUpdate`. ([#14134](https://github.com/expo/expo/pull/14134) by [@dani-mp](https://github.com/dani-mp))
```

Comment (dani-mp) [reply packages/expo-av/CHANGELOG.md:15]: It was probably my fault. I wanted to add the comma and missed the other change you proposed. 👍

**Review by bbarthec — COMMENTED** [review]
We're almost there, but we need to get rid of conflicts this PR is introducing 👀 
To to do so:
- you need to rebase with the current `master` branch (`git pull --rebase origin master`)
- then you need to rebuild the `build` directory in affected package (`cd packages/expo-av`, `yarn clean`, `yarn build`)
- add the resulting changes and then force push to this branch 💪

**Comment by dani-mp** [comment]
> We're almost there, but we need to get rid of conflicts this PR is introducing 👀
> To to do so:
> 
> * you need to rebase with the current `master` branch (`git pull --rebase origin master`)
> * then you need to rebuild the `build` directory in affected package (`cd packages/expo-av`, `yarn clean`, `yarn build`)
> * add the resulting changes and then force push to this branch 💪

Sorry, didn't have time yesterday to do it. Now it's done! Thanks again for your help.

**Comment by dani-mp** [comment]
Hey, @bbarthec! The PR has conflicts again :( Please, let me know when you'll have time to merge it so I can resolve the conflicts again.

**Comment by dani-mp** [comment]
Hi, @bbarthec. I've resolved the conflicts again! 🚀

File: apps/native-component-list/src/screens/AV/Player.tsx [hunk 9]
```diff
@@ -51,6 +52,7 @@ interface Props {
   shouldCorrectPitch: boolean;
   isPlaying: boolean;
   isMuted: boolean;
+  metadata: AVMetadata;
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/AV/Player.tsx:55]: We've got CI error on this one: https://github.com/expo/expo/pull/14134/checks?check_run_id=3505911371#step:7:6
We need to make it optional or adjust the VideoPlayer component that is also based on this class.
```suggestion
  metadata?: AVMetadata;
```

File: apps/native-component-list/src/screens/AV/Player.tsx [hunk 10]
```diff
@@ -175,6 +177,8 @@ export default function Player(props: Props) {
         {_renderReplayButton()}
       </View>
 
+      <Text>{props.metadata.title ?? ''}</Text>
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/AV/Player.tsx:180]:
```suggestion
      {props.metadata?.title && (<Text>{props.metadata.title}</Text>)}
```

**Review by bbarthec — CHANGES_REQUESTED** [review]
For some reason, I've failed to notice that the CI is failing, because of the one change you've made.
I've pointed it out in the comment 👇 
Adjust it and I believe that would be the last move of yours in this PR 😅

**Comment by dani-mp** [comment]
@barthap it's fixed now. Do you think the current failure is because of our changes? Could you re-run it?

**Comment by barthap** [comment]
I think you meant to mention `@bbarthec` 😅 
Regarding the CI failure - it's a common failure and pretty certainly not caused by you 😉

**Pull request merged**